File: src/healthcheck.ml
let socket_path =
  match Sys.getenv_opt "SOCKET_PATH" with
  | Some path -> path
  | None -> "/tmp/rinha-api.sock"

let request = "GET /ready HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\n\r\n"

let main () =
  let fd = Unix.socket Unix.PF_UNIX Unix.SOCK_STREAM 0 in
  Fun.protect
    ~finally:(fun () -> Unix.close fd)
    (fun () ->
      Unix.connect fd (Unix.ADDR_UNIX socket_path);
      ignore (Unix.write_substring fd request 0 (String.length request));
      let buf = Bytes.create 64 in
      let n = Unix.read fd buf 0 (Bytes.length buf) in
      if n >= 12 && String.equal (Bytes.sub_string buf 0 12) "HTTP/1.1 200"
      then 0
      else 1)

let () =
  try exit (main ()) with
  | _ -> exit 1

File: cpp/src/healthcheck.cpp
#include "common.hpp"

#include <cstdio>

int main() {
  const auto socket_path = rinha::env_string("SOCKET_PATH", "/tmp/rinha-api.sock");
  auto socket = rinha::connect_unix(socket_path);
  if (!socket) {
    std::perror("connect_unix");
    return 1;
  }

  constexpr std::string_view request =
      "GET /warmup HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\n\r\n";
  if (!rinha::write_all(socket.value, request.data(), request.size())) {
    return 1;
  }

  std::array<char, 128> buffer{};
  const auto n = ::read(socket.value, buffer.data(), buffer.size());
  if (n <= 0) {
    return 1;
  }
  const auto response = std::string_view{buffer.data(), static_cast<std::size_t>(n)};
  return response.starts_with("HTTP/1.1 200") ? 0 : 1;
}
